toolchain/base/value_store.h

```diff
@@ -84,8 +85,62 @@ namespace Internal {
 // Used as a parent class for non-printable types. This is just for
 // std::conditional, not as an API.
 class ValueStoreNotPrintable {};
+
+// Provides YAML printing of a value.
+template <typename ValueT>
+inline auto PrintValue(llvm::raw_ostream& out, const ValueT& val) {
+  out << val;
+};
+template <>
+inline auto PrintValue(llvm::raw_ostream& out, const llvm::APInt& val) {
+  val.print(out, /*isSigned=*/false);
+};
+template <>
+inline auto PrintValue(llvm::raw_ostream& out, const llvm::StringRef& val) {
+  out << "\"" << llvm::yaml::escape(val) << "\"";
+};
```

Comment:
OK, I see below what you're using this for. I think I would like to suggest:

- Turn these into overloads.
- Add a `DefaultPrinter` class:

```c++
struct DefaultPrinter {
  template<typename ValueT>
  void operator()(llvm::raw_ostream& out, const ValueT& value) {
    PrintValue(out, value);
  }
};
```

or if you prefer not to name the type:

```c++
inline constexpr auto default_printer =
    [](llvm::raw_ostream& out, const auto& value) {
      PrintValue(out, value);
    };
```

- Below, either pass in `DefaultPrinter()` (or the lambda) as the default:

```c++
template <typename ValueT>
inline auto PrintValueRange(
    // ...
    llvm::function_ref<void(llvm::raw_ostream&, const ValueT& val)> print =
        Internal::DefaultPrinter()) {
```

or allow a function object to be passed in and called directly (this is a bit more awkward if you use the lambda approach):

```c++
template <typename ValueT, typename PrintT = Internal::DefaultPrinter>
inline auto PrintValueRange(
    // ...
    PrintT print = {}) {
```

This would remove the need for the `std::decay_t` below, and allow use of overloads rather than function template specializations here, as well as permitting `PrintValue` to be an ADL extension point if we want.